spec/requests/groups_controller_spec.rb

```diff
@@ -858,6 +858,65 @@ def expect_type_to_return_right_groups(type, expected_group_ids)
 
       context "user default notifications" do
         it "should update default notification preference for existing users" do
+          user1 = Fabricate(:user)
+          user2 = Fabricate(:user)
+          group.add(user1)
+          group.add(user2)
+
+          put "/groups/#{group.id}.json", params: {
+            group: {
+              default_notification_level: 2
+            }
+          }
+
+          expect(response.status).to eq(200)
+          group_users = group.group_users
+          expect(response.parsed_body["user_count"]).to eq(group_users.count)
+
+          group_users.where(user: user1).update_all(notification_level: 1)
+
+          put "/groups/#{group.id}.json", params: {
+            group: {
+              default_notification_level: 2
+            }
+          }
+
+          expect(response.status).to eq(200)
+          expect(response.parsed_body["user_count"]).to eq(group.group_users.count - 1)
+
+          put "/groups/#{group.id}.json", params: {
+            group: {
+              default_notification_level: 2
+            },
+            update_existing_users: true
+          }
+
+          expect(response.status).to eq(200)
+          expect(response.parsed_body["success"]).to eq("OK")
+          expect(GroupUser.exists?(user: user2, group: group, notification_level: 2)).to be_truthy
+
+          put "/groups/#{group.id}.json", params: {
+            group: {
+              default_notification_level: 2
+            }
+          }
+
+          expect(response.status).to eq(200)
+          expect(response.parsed_body["success"]).to eq("OK")
+
+          put "/groups/#{group.id}.json", params: {
+            group: {
+              default_notification_level: 4
+            },
+            update_existing_users: false
+          }
+
+          expect(response.status).to eq(200)
+          expect(response.parsed_body["success"]).to eq("OK")
+          expect(GroupUser.exists?(user: user2, group: group, notification_level: 2)).to be_truthy
+        end
```

Comment:
```suggestion
          group.update!(default_notification_level: NotificationLevels.all[:watching])
          user1 = Fabricate(:user)
          user2 = Fabricate(:user)
          group.add(user1)
          group.add(user2)
          group_user1 = user1.group_users.first
          group_user2 = user2.group_users.first

          put "/groups/#{group.id}.json", params: {
            group: {
              default_notification_level: NotificationLevels.all[:tracking]
            }
          }

          expect(response.status).to eq(200)

          expect(group_user1.reload.notification_level).to eq(NotificationLevels.all[:watching])
          expect(group_user2.reload.notification_level).to eq(NotificationLevels.all[:watching])

          group_users = group.group_users
          expect(response.parsed_body["user_count"]).to eq(group_users.count)

          group_user1.update!(notification_level: NotificationLevels.all[:regular])

          put "/groups/#{group.id}.json", params: {
            group: {
              default_notification_level: NotificationLevels.all[:tracking]
            }
          }

          expect(response.status).to eq(200)
          expect(response.parsed_body["user_count"]).to eq(group.group_users.count - 1)
          expect(group_user1.reload.notification_level).to eq(NotificationLevels.all[:watching])
          expect(group_user2.reload.notification_level).to eq(NotificationLevels.all[:watching])

          put "/groups/#{group.id}.json", params: {
            group: {
              default_notification_level: NotificationLevels.all[:tracking]
            },
            update_existing_users: true
          }

          expect(response.status).to eq(200)
          expect(response.parsed_body["success"]).to eq("OK")
          expect(group_user1.reload.notification_level).to eq(NotificationLevels.all[:tracking])
          expect(group_user2.reload.notification_level).to eq(NotificationLevels.all[:tracking])

          put "/groups/#{group.id}.json", params: {
            group: {
              default_notification_level: NotificationLevels.all[:regular]
            },
            update_existing_users: false
          }

          expect(response.status).to eq(200)
          expect(response.parsed_body["success"]).to eq("OK")
          expect(group_user1.reload.notification_level).to eq(NotificationLevels.all[:tracking])
          expect(group_user2.reload.notification_level).to eq(NotificationLevels.all[:tracking])
        end
```

Reply:
@vinothkannans I wrote up how I would go about writing the tests which asserts for the state of the resources which we're interested in. In particular, I think the difference in the response when a user has been updated vs when a user has not been updated is confusing and I hope we can fix that at some point.